crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -92,23 +94,40 @@ impl CommandQueue {
         // flush the previously queued entities
         world.flush();
 
+        // Pointer that will iterate over the entries of the buffer.
+        let mut cursor = self.bytes.as_mut_ptr();
+
+        // The address of the end of the buffer.
+        let end_addr = cursor as usize + self.bytes.len();
+
         // Reset the buffer, so it can be reused after this function ends.
         // In the loop below, ownership of each command will be transferred into user code.
         // SAFETY: `set_len(0)` is always valid.
         unsafe { self.bytes.set_len(0) };
 
-        for meta in self.metas.drain(..) {
-            // SAFETY: `CommandQueue` guarantees that each metadata must have a corresponding value stored in `self.bytes`,
-            // so this addition will not overflow its original allocation.
-            let cmd = unsafe { self.bytes.as_mut_ptr().add(meta.offset) };
+        while (cursor as usize) < end_addr {
+            // SAFETY: The cursor is either at the start of the buffer, or just after the previous command.
+            // Since we know that the cursor is in bounds, it must point to the start of a new command.
+            let meta = unsafe { cursor.cast::<CommandMeta>().read_unaligned() };
+            // Advance to the bytes just after `meta`, which represent a type-erased command.
+            // SAFETY: For most types of `Command`, the pointer immediately following the metadata
+            // is guaranteed to be in bounds. If the command is a zero-sized type (ZST), then the cursor
+            // might be 1 byte past the end of the buffer, which is safe.
+            cursor = unsafe { cursor.add(std::mem::size_of::<CommandMeta>()) };
+            // Construct an owned pointer to the command.
             // SAFETY: It is safe to transfer ownership out of `self.bytes`, since the call to `set_len(0)` above
             // gaurantees that nothing stored in the buffer will get observed after this function ends.
             // `cmd` points to a valid address of a stored command, so it must be non-null.
-            let cmd = unsafe { OwningPtr::new(NonNull::new_unchecked(cmd.cast())) };
-            // SAFETY: The underlying type of `cmd` matches the type expected by `meta.apply_command`.
-            unsafe {
-                (meta.apply_command)(cmd, world);
-            }
+            let cmd = unsafe { OwningPtr::new(std::ptr::NonNull::new_unchecked(cursor.cast())) };
```

Comment:
This may not conform to the safety invariant that OwningPtr must be aligned. Even if it is, it doesn't seem to be documented in the safety comment.

Reply:
This creates an `OwningPtr<Unaligned>`, so there is no alignment requirement.

Reply:
Ah brainfart, it is required by the type signature of the function. Could we make that explicit here and ensure it's covered by the safety comment?

Reply:
i just added the turbofish operator to make it more obvious that this is correct.